core.js

```diff
@@ -60,16 +44,120 @@ function _check(buffer, headers, options) {
 }
 
 export async function fileTypeFromTokenizer(tokenizer) {
-	try {
-		return new FileTypeParser().parse(tokenizer);
-	} catch (error) {
-		if (!(error instanceof strtok3.EndOfStreamError)) {
-			throw error;
-		}
+	return new FileTypeParser().fromTokenizer(tokenizer);
+}
+
+export class TokenizerPositionError extends Error {
+	constructor(message = 'Tokenizer position changed without detecting a file type.') {
+		super(message);
+		this.name = 'TokenizerPositionError';
 	}
 }
 
-class FileTypeParser {
+export class FileTypeParser {
+	constructor(options) {
+		this.detectors = options?.customDetectors;
+
+		this.runCustomDetectors = this.#runCustomDetectors.bind(this);
+		this.fromTokenizer = this.fromTokenizer.bind(this);
+		this.fromBuffer = this.fromBuffer.bind(this);
+		this.parse = this.parse.bind(this);
+	}
+
+	async #runCustomDetectors(tokenizer) {
+		const {detectors} = this;
+		const initialPosition = tokenizer?.position;
+
+		for (const detector of detectors || []) {
+			const fileType = await detector(tokenizer);
+			if (fileType) {
+				return fileType;
+			}
+
+			if (initialPosition !== tokenizer?.position) {
+				throw new TokenizerPositionError();
+			}
+		}
+
+		return undefined;
+	}
+
+	async fromTokenizer(tokenizer) {
+		const {parse, runCustomDetectors} = this;
+		try {
+			const customFileType = await runCustomDetectors(tokenizer);
```

Comment:
You miss the check here if the tokenizer has been read.

You could consider to call parse from `#runCustomDetectors(tokenizer)`, after the for loop

Reply:
I'm sorry, I don't see in what scenario the check is missed. It happens inside the `runCustomDetectors(...)`, and I think that's where it belongs, as it's part of the custom detectors functionality.

Reply:
How do you prevent internal detectors from running when a custom detector has consumed the stream?

You have now protected the remaing detectors when undefined & tokenizer has been read, situation occured, but not the internal detectors.

Reply:
Sorry I missed the part that you further on were catching this error, it's something I did not directly expect.

I made some suggestions in a branch, how I believe the same thing can be achieved, simpler, with less code: https://github.com/sindresorhus/file-type/tree/extend-file-type-updated-suggested-code-changes

Reply:
Thank you very much for making the effort. Looks okay to me, just merged it.